src/useMergedRefs.ts

```diff
@@ -1,22 +1,42 @@
-import { useMemo } from 'react'
+import { RefCallback, useMemo, version } from 'react'
+
+const isReact19 = parseInt(version.split('.')[0]!, 10) >= 19
 
 type CallbackRef<T> = (ref: T | null) => void
 type Ref<T> = React.MutableRefObject<T> | CallbackRef<T>
 
 function toFnRef<T>(ref?: Ref<T> | null) {
-  return !ref || typeof ref === 'function'
-    ? ref
-    : (value: T | null) => {
-        ref.current = value as T
-      }
+  if (!ref || typeof ref === 'function') {
+    return ref
+  }
+
+  return (value: T | null) => {
+    ref.current = value as T
+  }
+}
+
+function cleanUp<T>(cleanup: unknown, ref: RefCallback<T> | null | undefined) {
+  if (typeof cleanup === 'function') {
+    cleanup()
+  } else if (ref) {
+    ref(null)
+  }
 }
 
 export function mergeRefs<T>(refA?: Ref<T> | null, refB?: Ref<T> | null) {
-  const a = toFnRef(refA)
-  const b = toFnRef(refB)
-  return (value: T | null) => {
-    if (a) a(value)
-    if (b) b(value)
+  const refASetter = toFnRef(refA)
+  const refBSetter = toFnRef(refB)
+
+  if (isReact19) {
+    return (value: T | null) => {
+      const cleanupA = refASetter?.(value)
+      const cleanupB = refBSetter?.(value)
+
+      return () => {
+        cleanUp(cleanupA, refASetter)
+        cleanUp(cleanupB, refBSetter)
+      }
+    }
```

Comment:
```suggestion
  return (value: T | null) => {
    const cleanupA = refASetter?.(value)
    const cleanupB = refBSetter?.(value)

    if (isReact19) {
      return () => {
        cleanUp(cleanupA, refASetter)
        cleanUp(cleanupB, refBSetter)
      }
    }
```

Moved R19 check inside handler as the generated setters aren't called on R18.  Sorry, formatting might be messed since I'm doing this from the web UI